iceberg_rust_ffi/src/full.rs

```diff
@@ -0,0 +1,80 @@
+use std::ffi::{c_char, c_void, CStr};
+use std::ptr;
+
+use iceberg::scan::{TableScan, TableScanBuilder};
+use object_store_ffi::{
+    export_runtime_op, with_cancellation, CResult, NotifyGuard, ResponseGuard, RT,
+};
+use tokio::sync::Mutex as AsyncMutex;
+
+use crate::scan_common::*;
+use crate::{IcebergArrowStream, IcebergArrowStreamResponse, IcebergTable};
+
+/// Struct for regular (full) scan builder and scan
+#[repr(C)]
+pub struct IcebergScan {
+    pub builder: Option<TableScanBuilder<'static>>,
+    pub scan: Option<TableScan>,
+}
+
+unsafe impl Send for IcebergScan {}
+
+/// Create a new scan builder
+#[no_mangle]
+pub extern "C" fn iceberg_new_scan(table: *mut IcebergTable) -> *mut IcebergScan {
+    if table.is_null() {
+        return ptr::null_mut();
+    }
+    let table_ref = unsafe { &*table };
+    let scan_builder = table_ref.table.scan();
+    Box::into_raw(Box::new(IcebergScan {
+        builder: Some(scan_builder),
+        scan: None,
+    }))
+}
+
+// Use macros from scan_common for shared functionality
+impl_select_columns!(iceberg_select_columns, IcebergScan);
```

Comment:
In this case, I think it is fine to use macros. However, in general macros should also be sometimes avoided, so we could go instead with traits for the builder + scan, and a generic `ScanBase<B: BuilderTrait, S: ScanTrait>` variant, if we want to keep it more extensible.

Reply:
Ok, agree it's better not to rely on macros in future, thanks for pointing out.